File: app/src/main/java/com/example/softmusic/songList/MusicSongListAdapter.java
package com.example.softmusic.songList;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import com.example.softmusic.databinding.CardSongListBinding;
import com.example.softmusic.R;

import java.util.List;

public class MusicSongListAdapter extends RecyclerView.Adapter<MusicSongListAdapter.ViewHolder> {

    private static final String TAG = "MusicSongListAdapter";
    
    private List<MusicSongList> musicSongListList;
    private final Context context;

    public MusicSongListAdapter(Context context, List<MusicSongList> musicSongListList) {
        this.musicSongListList = musicSongListList;
        this.context = context;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        CardSongListBinding cardSongListBinding = DataBindingUtil.inflate(
                (LayoutInflater.from(context)), R.layout.card_song_list,parent,false);
        return new ViewHolder(cardSongListBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, @SuppressLint("RecyclerView") int position) {
        holder.getCardSongListBinding().setMusicListItem(musicSongListList.get(position));
        int[] array = {R.drawable.card,R.drawable.card2,R.drawable.card3,R.drawable.card3};
        holder.getCardSongListBinding().imageSongList.setBackgroundResource(array[position]);
        holder.getCardSongListBinding().songListItem.setOnClickListener(view -> {
            NavController controller = Navigation.findNavController(view);
            // TODO 传递Bundle（Room中的索引）
            Bundle bundle = new Bundle();
            bundle.putString("key",musicSongListList.get(position).getSongListTitle());
            Log.d(TAG, "onBindViewHolder: " + musicSongListList.get(position).getSongListTitle());
            controller.navigate(R.id.action_musicSongListFragment2_to_musicSongFragment2,bundle);
        });
    }

    @Override
    public int getItemCount() {
        return musicSongListList.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder{
        CardSongListBinding cardSongListBinding;
        public ViewHolder(@NonNull CardSongListBinding cardSongListBinding) {
            super(cardSongListBinding.getRoot());
            this.cardSongListBinding = cardSongListBinding;
        }

        public CardSongListBinding getCardSongListBinding() {
            return cardSongListBinding;
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setMusicSongListList(List<MusicSongList> musicSongListList) {
        this.musicSongListList = musicSongListList;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/softmusic/songList/MusicSongListAdapter.kt
package com.example.softmusic.songList

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.AttrRes
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import androidx.navigation.NavController
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import com.example.softmusic.R
import com.example.softmusic.databinding.CardSongListBinding
import com.example.softmusic.entity.MusicSongList

class MusicSongListAdapter(
    private val context: Context,
    private var musicSongListList: List<MusicSongList>,
    private var selectId:Long
) : RecyclerView.Adapter<MusicSongListAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val cardSongListBinding: CardSongListBinding = CardSongListBinding.inflate(LayoutInflater.from(context),parent,false)
        return ViewHolder(cardSongListBinding)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: ViewHolder, @SuppressLint("RecyclerView") position: Int) {
        with(holder.cardSongListBinding)
        {
            songNumber.text = musicSongListList[position].songNumber.toString() + '首'
            songListTitle.text = musicSongListList[position].songListTitle
            val array =
                intArrayOf(R.drawable.card, R.drawable.card2, R.drawable.card3, R.drawable.card3)
            imageSongList.setBackgroundResource(array[0])
            if (musicSongListList[position].musicSongListId == selectId){
                songListTitle.setTextColor(context.resolveColorAttr(androidx.appcompat.R.attr.colorPrimaryDark))
                songNumber.setTextColor(context.resolveColorAttr(androidx.appcompat.R.attr.colorPrimaryDark))
            } else {
                songListTitle.setTextColor(context.resolveColorAttr(androidx.appcompat.R.attr.colorAccent))
                songNumber.setTextColor(context.resolveColorAttr(androidx.appcompat.R.attr.colorAccent))
            }
            songListItem.setOnClickListener { view: View ->
                val controller: NavController = findNavController(view)
                val bundle = Bundle()
                bundle.putLong("key", musicSongListList[position].musicSongListId)
                controller.navigate(R.id.action_to_song_fragment, bundle)
            }
        }

    }

    override fun getItemCount(): Int {
        return musicSongListList.size
    }

    class ViewHolder(var cardSongListBinding: CardSongListBinding) : RecyclerView.ViewHolder(
        cardSongListBinding.root
    )

    @SuppressLint("NotifyDataSetChanged")
    fun setMusicSongListList(musicSongListList: List<MusicSongList>) {
        this.musicSongListList = musicSongListList
        notifyDataSetChanged()
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setSelectedId(id:Long){
        selectId = id
        notifyDataSetChanged()
    }

    @ColorInt
    private fun Context.resolveColorAttr(@AttrRes colorAttr: Int): Int {
        val resolvedAttr = resolveThemeAttr(colorAttr)
        val colorRes = if (resolvedAttr.resourceId != 0) resolvedAttr.resourceId else resolvedAttr.data
        return ContextCompat.getColor(this, colorRes)
    }

    private fun Context.resolveThemeAttr(@AttrRes attrRes: Int): TypedValue {
        val typedValue = TypedValue()
        theme.resolveAttribute(attrRes, typedValue, true)
        return typedValue
    }
}